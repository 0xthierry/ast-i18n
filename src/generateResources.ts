import fs from 'graceful-fs';
import * as babel from '@babel/core';
import BabelPluginI18n from './BabelPluginI18n';

import babelConfig from '../babel.config.js';

const resource = (i18nResource: {[key: string]: string}) => {
  const formatted = Object.keys(i18nResource)
    .map(key => `   '${key}': \`${i18nResource[key]}\``)
    .join(',\n');

  return `export default {
  translation: {
  ${formatted}
  }
}
`
};

export const generateResources = (files: string[]) => {
  let phrases = [];
  for (const filename of files) {
    const source = fs.readFileSync(filename, 'utf8');

    try {
      babel.transformSync(source, {
        ast: false,
        code: true,
        plugins: [...babelConfig.plugins, BabelPluginI18n],
        // TODO - add StringExtractPlugin
        // plugins: SyntaxPlugins.list.concat([[fbt, options]]),
        sourceType: 'unambiguous',
        filename,
      });

      const newPhrases = BabelPluginI18n.getExtractedStrings();

      phrases = [
        ...phrases,
        ...newPhrases,
      ];
    } catch (err) {
      console.log('err: ', filename, err);
    }
  }

  const i18nMap = BabelPluginI18n.getI18Map();

  fs.writeFileSync('resource.tsx', resource(i18nMap));

  console.log('all collected texts');
  console.log(phrases);
};
