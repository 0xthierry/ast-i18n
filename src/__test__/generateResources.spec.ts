import path from "path";

import { generateResources } from '../generateResources';

const defineTest = (dirName: string, testFilePrefix: string) => {
  const testName = `extra string from ${testFilePrefix}`;

  it(testName, () => {
    const fixtureDir = path.join(dirName, '..', '__testfixtures__');
    const inputPath = path.join(fixtureDir, testFilePrefix + '.input.tsx');

    const files = [inputPath];
    const i18nMap = generateResources(files);

    expect(i18nMap).toMatchSnapshot();
  });
};

describe('generateResources', () => {
  defineTest(__dirname, 'Classes');
});
