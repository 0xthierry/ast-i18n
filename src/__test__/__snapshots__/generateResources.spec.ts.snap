// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateResources extra string from Classes 1`] = `
Object {
  "my_great_class_component": "my great class component",
}
`;
